### PR TITLE
Add `event_stream_ordering` column to membership state tables

### PR DESCRIPTION
**Pull request by Fizzadar**
This adds an `event_stream_ordering` column to `current_state_events`, `local_current_membership` and `room_memberships`. Each of these tables is regularly joined with the `events` table to get the stream ordering and denormalising this into each table will yield significant query performance improvements once used. Includes a background job to populate these values from the `events` table.

Same idea as https://github.com/matrix-org/synapse/pull/13703 which has been a significant improvement on our database instance, suspect once applied this will be far greater.

Signed off by Nick @ Beeper (@fizzadar).

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog).
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

**Review by squahtx — APPROVED** [review]
lgtm, excited to see the coming performance improvements.

File: synapse/storage/databases/main/events_worker.py [hunk 1]
```diff
@@ -1791,10 +1791,10 @@ def get_ex_outlier_stream_rows_txn(
                 " LEFT JOIN event_relations USING (event_id)"
                 " LEFT JOIN room_memberships USING (event_id)"
                 " LEFT JOIN rejections USING (event_id)"
-                " WHERE ? < event_stream_ordering"
-                " AND event_stream_ordering <= ?"
+                " WHERE ? < out.event_stream_ordering"
+                " AND out.event_stream_ordering <= ?"
                 " AND out.instance_name = ?"
-                " ORDER BY event_stream_ordering ASC"
+                " ORDER BY out.event_stream_ordering ASC"
```
Comment (DMRobertson) [inline synapse/storage/databases/main/events_worker.py:1797]: This disambiguates the query. Unfortunately it means that old old Synapses (without this code change) will fail to run this query against an upgraded database.

Not sure how we should handle this.

File: synapse/storage/schema/main/delta/73/26membership_tables_event_stream_ordering.sql [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+ALTER TABLE current_state_events ADD COLUMN event_stream_ordering BIGINT;
+ALTER TABLE local_current_membership ADD COLUMN event_stream_ordering BIGINT;
+ALTER TABLE room_memberships ADD COLUMN event_stream_ordering BIGINT;
+
+INSERT INTO background_updates (update_name, progress_json) VALUES
+  ('populate_membership_event_stream_ordering', '{}');
```
Comment (clokep) [inline synapse/storage/schema/main/delta/73/26membership_tables_event_stream_ordering.sql:21]: This probably should have defined an `ordering`, right now it jumped ahead of all the other pending updates which seems not great.

File: synapse/storage/databases/main/events_worker.py (hunk 1 above)
Comment (DMRobertson) [reply synapse/storage/databases/main/events_worker.py:1797]: https://github.com/matrix-org/synapse/issues/15014

**Comment by clokep** [comment]
FYI This was reverted, see #15014.

**Pull request merged**